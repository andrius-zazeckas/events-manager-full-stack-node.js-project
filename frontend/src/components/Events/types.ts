export type TEvent = {
  id: number;
  event_name: string | null;
  description: string | null;
  date: string | number | Date;
};

export type TEventVisitors = {
  id: number;
  full_name: string | null;
  email: string | null;
  age: number | null;
  date_of_birth: string | number | Date;
};

export type TEventProps = { event: TEvent };
export type TEventVisitorProps = { eventVisitor: TEventVisitors };
