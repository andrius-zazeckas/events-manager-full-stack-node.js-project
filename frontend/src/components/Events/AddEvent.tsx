import {
  Box,
  Button,
  FormControl,
  InputLabel,
  OutlinedInput,
  Typography,
} from "@mui/material";
import axios from "axios";
import { ChangeEvent, FormEventHandler, useState } from "react";
import { useNavigate } from "react-router-dom";
import type { TEvent } from "./types";

export const AddEvent = () => {
  const [newEvent, setNewEvent] = useState<TEvent>({} as TEvent);

  const navigate = useNavigate();

  const handleInputChange = (
    event: ChangeEvent<HTMLInputElement | HTMLTextAreaElement>,
    prop: string
  ) => {
    setNewEvent({
      ...newEvent,
      [prop]: event.target.value,
    });

    console.log(newEvent);
  };

  const resetForm = () => {
    setNewEvent({} as TEvent);
  };

  const handleSubmit: FormEventHandler<HTMLFormElement> = (event) => {
    event.preventDefault();

    axios
      .post(
        `http://localhost:5000/events/add-event`,
        {
          event_name: newEvent.event_name,
          date: newEvent.date,
          description: newEvent.description,
        },
        {
          headers: { authorization: `Bearer ${localStorage.getItem("token")}` },
        }
      )
      .then(() => {
        alert(`Event added successfully`);

        resetForm();
        navigate(-1);
      })
      .catch((error) => {
        //   alert(error.response.data.error);
        console.error(error.response.data.error);
      });
  };

  return (
    <Box margin="0 auto">
      <Box>
        <Box>
          <Typography
            textAlign="center"
            variant="h2"
            margin="20px"
            fontSize="40px"
          >
            Register new User
          </Typography>
        </Box>
        <Box
          component="form"
          display="grid"
          maxWidth="300px"
          gap="10px"
          mx="auto"
          my="40px"
          onSubmit={handleSubmit}
        >
          <FormControl>
            <InputLabel htmlFor="Event name">Event name</InputLabel>
            <OutlinedInput
              label="Event name"
              required
              autoFocus
              value={newEvent.event_name ?? ""}
              onChange={(event) => handleInputChange(event, "event_name")}
            />
          </FormControl>

          <FormControl>
            <InputLabel htmlFor="Description">Description</InputLabel>
            <OutlinedInput
              label="Description"
              required
              value={newEvent.description ?? ""}
              onChange={(event) => handleInputChange(event, "description")}
            />
          </FormControl>

          <FormControl>
            <InputLabel shrink htmlFor="date">
              Event date
            </InputLabel>
            <OutlinedInput
              notched
              type="date"
              label="Event date"
              required
              value={newEvent.date ?? ""}
              onChange={(event) => handleInputChange(event, "date")}
            />
          </FormControl>

          <Box display="flex" justifyContent="center" gap="20px">
            <Button variant="outlined" onClick={() => navigate(-1)}>
              Cancel
            </Button>
            <Button type="submit" variant="contained">
              Submit
            </Button>
          </Box>
        </Box>
      </Box>
    </Box>
  );
};
