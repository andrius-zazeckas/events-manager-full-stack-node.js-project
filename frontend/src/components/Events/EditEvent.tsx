import { Box, Button, FormControl, TextField, Typography } from "@mui/material";
import axios from "axios";
import {
  ChangeEvent,
  type FC,
  FormEventHandler,
  useEffect,
  useState,
} from "react";
import { useNavigate, useParams } from "react-router-dom";
import type { TEvent } from "./types";

export const EditEvent: FC = () => {
  const [tempEvent, setTempEvent] = useState<TEvent>({} as TEvent);
  const [isLoading, setIsLoading] = useState(true);
  const [updatedEvent, setUpdatedEvent] = useState<{
    [key: string]: string;
  }>({});

  const formatedEventDate = new Date(tempEvent?.date).toLocaleDateString(
    "lt-LT"
  );

  const params = useParams();
  const navigate = useNavigate();

  useEffect(() => {
    axios
      .get(`http://localhost:5000/events/event/${params.id}`, {
        headers: {
          authorization: `Bearer ${localStorage.getItem("token")}`,
        },
      })
      .then((res) => {
        if (Array.isArray(res.data)) {
          setTempEvent({
            event_name: res.data[0]?.event_name,
            description: res.data[0]?.description,
            date: res.data[0]?.date,
            image: res.data[0]?.image,
          } as TEvent);
        }
      })
      .finally(() => {
        setTimeout(() => {
          setIsLoading(false);
        }, 1000);
      })
      .catch((error) => {
        console.error(error);
        alert(error.response.data.error);
      });
  }, [params.id, setTempEvent]);

  const handleInputChange = (
    event: ChangeEvent<HTMLInputElement | HTMLTextAreaElement>,
    prop: string
  ) => {
    setUpdatedEvent({
      ...updatedEvent,
      [prop]: event.target.value,
    });
  };

  const resetForm = () => {
    setUpdatedEvent({});
  };

  const handleSubmit: FormEventHandler<HTMLFormElement> = (event) => {
    event.preventDefault();

    const formatedUpdatedEventDate = new Date(
      updatedEvent?.date
    ).toLocaleDateString("lt-LT");

    if (!Object.keys(updatedEvent).length) {
      return alert("Please update the form before submitting");
    }

    if (window.confirm("Are you sure you want to edit this event?")) {
      axios
        .patch(
          `http://localhost:5000/events/edit-event/${params.id}`,
          {
            event_name: updatedEvent?.event_name
              ? updatedEvent?.event_name
              : tempEvent?.event_name,
            description: updatedEvent?.description
              ? updatedEvent?.description
              : tempEvent?.description,
            date: updatedEvent?.date
              ? formatedUpdatedEventDate
              : formatedEventDate,
            image: updatedEvent?.image ? updatedEvent?.image : tempEvent?.image,
          },
          {
            headers: {
              authorization: `Bearer ${localStorage.getItem("token")}`,
            },
          }
        )
        .then(() => {
          alert(`Event updated successfully`);

          resetForm();

          navigate(-1);
        })
        .catch((error) => {
          alert(error.response.data.error);
          console.error(error.response.data.error);
        });
    }
  };

  return (
    <Box textAlign="center">
      {isLoading ? (
        <Typography variant="h3">Loading...</Typography>
      ) : (
        <Box>
          <Box>
            <Typography
              textAlign="center"
              variant="h2"
              margin="20px"
              fontSize="30px"
            >
              Edit Event
            </Typography>
          </Box>

          <Box
            component="form"
            display="grid"
            maxWidth="300px"
            gap="10px"
            mx="auto"
            my="40px"
            onSubmit={handleSubmit}
          >
            <FormControl>
              <TextField
                id="edit-event-name"
                aria-label="edit event name"
                label="Event name"
                variant="outlined"
                required
                defaultValue={tempEvent?.event_name}
                onChange={(event) => handleInputChange(event, "event_name")}
              />
            </FormControl>

            <FormControl>
              <TextField
                id="edit-event-description"
                aria-label="edit event description"
                label="Description"
                variant="outlined"
                required
                defaultValue={tempEvent?.description}
                onChange={(event) => handleInputChange(event, "description")}
              />
            </FormControl>

            <FormControl>
              <TextField
                id="edit-event-image"
                aria-label="edit event image"
                label="Image URL"
                variant="outlined"
                required
                defaultValue={tempEvent?.image}
                onChange={(event) => handleInputChange(event, "image")}
              />
            </FormControl>

            <FormControl>
              <TextField
                id="edit-date"
                aria-label="edit date"
                label="Event date"
                type="date"
                InputLabelProps={{ shrink: true }}
                variant="outlined"
                required
                defaultValue={formatedEventDate}
                onChange={(event) => handleInputChange(event, "date")}
              />
            </FormControl>

            <Box display="flex" justifyContent="center" gap="20px">
              <Button variant="outlined" onClick={() => navigate(-1)}>
                Cancel
              </Button>
              <Button type="submit" variant="contained">
                Submit
              </Button>
            </Box>
          </Box>
        </Box>
      )}
    </Box>
  );
};
