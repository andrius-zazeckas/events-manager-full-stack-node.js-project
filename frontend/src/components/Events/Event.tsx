import { Box, Button, Grid, Typography } from "@mui/material";
import axios from "axios";
import { type FC, useContext } from "react";
import { useNavigate } from "react-router-dom";
import { EventsContext } from "../Contexts/EventsContext";
import { TEventProps } from "./types";

export const Event: FC<TEventProps> = ({ event }) => {
  const { events, setEvents } = useContext(EventsContext);
  const formatedDate = new Date(event.date).toLocaleDateString("lt-LT");

  const navigate = useNavigate();

  const handleClick = () => {
    navigate(`/events/event-visitors/${event.id}`);
  };

  const handleEditClick = () => {
    window.location.assign(`/events/edit-event/${event.id}`);
  };

  const handleDeleteClick = () => {
    if (window.confirm("Are you sure you want to delete this event?")) {
      axios
        .delete(`http://localhost:5000/events/delete-event/${event.id}`, {
          headers: {
            authorization: `Bearer ${localStorage.getItem("token")}`,
          },
        })
        .then(() => {
          alert(`Event deleted successfully`);

          setEvents((prevEvent) =>
            prevEvent.filter((deletedEvent) => deletedEvent.id !== event.id)
          );
        })
        .catch((error) => {
          alert(error.response.data.error);
          console.error(error.response.data.error);
        });
    }
  };

  return (
    <Grid
      item
      xs={6}
      sm={6}
      md={3}
      padding="15px"
      boxShadow="0px 0px 8px 1px rgba(0, 0, 0, 0.1)"
      borderRadius="5px"
      textAlign="center"
      sx={{
        cursor: "pointer",
        ":hover": { bgcolor: "#DFE6E9", color: "#2D3436" },
      }}
    >
      <Box onClick={handleClick}>
        <Typography variant="h4" fontSize="20px" fontWeight="550" my="20px">
          {event.event_name}
        </Typography>
        <Box
          display="flex"
          margin="0 auto"
          width="100%"
          height="200px"
          alignItems="center"
          aria-label="event image"
          sx={{
            "& img": { objectFit: "cover", width: "100%", maxHeight: "100%" },
          }}
          borderRadius="5px"
        >
          <img
            src={event && event.image}
            alt={(event && event.event_name) ?? "event image"}
          />
        </Box>
        <Box my="10px">
          <Typography component="p" fontSize="15px">
            {event.description}
          </Typography>
          <Typography>{formatedDate}</Typography>
        </Box>
      </Box>

      <Box display="flex" justifyContent="center" gap="20px">
        <Button variant="outlined" onClick={handleEditClick}>
          Edit
        </Button>
        <Button
          variant="outlined"
          color="secondary"
          onClick={handleDeleteClick}
        >
          Delete
        </Button>
      </Box>
    </Grid>
  );
};
