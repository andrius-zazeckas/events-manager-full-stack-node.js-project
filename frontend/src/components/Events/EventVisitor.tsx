import { FC, useContext } from "react";
import type { TEventVisitorProps } from "./types";
import { Button, TableCell, TableRow } from "@mui/material";
import axios from "axios";
import { EventsContext } from "../Contexts/EventsContext";

export const EventVisitor: FC<TEventVisitorProps> = ({ eventVisitor }) => {
  const formatedDate = new Date(eventVisitor.date_of_birth).toLocaleDateString(
    "lt-LT"
  );

  const { setEventVisitors } = useContext(EventsContext);

  const full_name = `${eventVisitor.first_name} ${eventVisitor.last_name}`;

  const handleEditClick = () => {
    window.location.assign(`/visitors/edit-visitor/${eventVisitor.id}`);
  };

  const handleDeleteClick = () => {
    if (window.confirm("Are you sure you want to delete this visitor?")) {
      axios
        .delete(
          `http://localhost:5000/visitors/delete-visitor/${eventVisitor.id}`,
          {
            headers: {
              authorization: `Bearer ${localStorage.getItem("token")}`,
            },
          }
        )
        .then(() => {
          alert(`Visitor deleted successfully`);

          setEventVisitors((prevVisitor) =>
            prevVisitor.filter(
              (deletedVisitor) => deletedVisitor.id !== eventVisitor.id
            )
          );
        })
        .catch((error) => {
          alert(error.response.data.error);
          console.error(error.response.data.error);
        });
    }
  };

  return (
    <TableRow sx={{ "&:last-child td, &:last-child th": { border: 0 } }}>
      <TableCell align="center">{eventVisitor.id}</TableCell>
      <TableCell align="center">{full_name}</TableCell>
      <TableCell align="center">{eventVisitor.email}</TableCell>
      <TableCell align="center">{formatedDate}</TableCell>
      <TableCell align="center">{eventVisitor.age}</TableCell>
      <TableCell align="right">
        <Button onClick={handleEditClick}>Edit</Button>
      </TableCell>
      <TableCell align="left">
        <Button onClick={handleDeleteClick}>Delete</Button>
      </TableCell>
    </TableRow>
  );
};
