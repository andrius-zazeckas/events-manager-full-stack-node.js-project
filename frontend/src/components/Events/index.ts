export { Event } from "./Event";
export { Events } from "./Events";
export { EventVisitor } from "./EventVisitor";
export { EventVisitors } from "./EventVisitors";
export { AddEvent } from "./AddEvent";
