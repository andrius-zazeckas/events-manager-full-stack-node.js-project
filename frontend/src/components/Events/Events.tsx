import { Box, Grid, Typography } from "@mui/material";
import axios from "axios";
import { useContext, useEffect, useState } from "react";
import { EventsContext } from "../Contexts/EventsContext";
import { Event } from "./Event";

export const Events = () => {
  const { events, setEvents } = useContext(EventsContext);
  const [isLoading, setIsLoading] = useState(false);

  const getEvents = () => {
    axios
      .get("http://localhost:5000/events", {
        headers: { authorization: `Bearer ${localStorage.getItem("token")}` },
      })
      .then((res) => {
        if (Array.isArray(res.data)) {
          setEvents(res.data);
        }
      })
      .catch((error) => {
        console.error(error);
        // alert(error.response.data.error);
      })
      .finally(() => {
        setTimeout(() => {
          setIsLoading(false);
        }, 1000);
      });
  };

  useEffect(() => {
    setIsLoading(true);
    getEvents();
  }, []);

  console.log(events);

  return (
    <Box display="flex" textAlign="center" justifyContent="center">
      {isLoading ? (
        <Box margin="40px">
          <Typography variant="h3">Loading...</Typography>
        </Box>
      ) : (
        <Grid container gap="20px" marginTop="40px" justifyContent="center">
          {events.map((event) => (
            <Event key={event.id} event={event} />
          ))}
        </Grid>
      )}
    </Box>
  );
};
