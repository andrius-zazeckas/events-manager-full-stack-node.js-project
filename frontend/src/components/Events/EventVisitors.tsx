import {
  Box,
  Paper,
  Table,
  TableBody,
  TableCell,
  TableContainer,
  TableHead,
  TableRow,
  Typography,
} from "@mui/material";
import axios from "axios";
import { useContext, useEffect, useState } from "react";
import { useParams } from "react-router-dom";
import { EventsContext } from "../Contexts/EventsContext";
import { EventVisitor } from "./EventVisitor";

export const EventVisitors = () => {
  const { eventVisitors, setEventVisitors } = useContext(EventsContext);
  const [isLoading, setIsLoading] = useState(false);

  const params = useParams();

  useEffect(() => {
    setIsLoading(true);

    axios
      .get(`http://localhost:5000/events/event-visitors/${params.id}`, {
        headers: { authorization: `Bearer ${localStorage.getItem("token")}` },
      })
      .then((res) => {
        if (Array.isArray(res.data)) {
          setEventVisitors(res.data);
        }
      })
      .catch((error) => {
        console.error(error);
        // alert(error.response.data.error);
      })
      .finally(() => {
        setTimeout(() => {
          setIsLoading(false);
        }, 1000);
      });
  }, [params.id, setEventVisitors]);

  return (
    <Box
      display="flex"
      textAlign="center"
      justifyContent="center"
      //   maxWidth="800px"
      mx="auto"
    >
      {isLoading ? (
        <Box margin="40px">
          <Typography variant="h3">Loading...</Typography>
        </Box>
      ) : (
        <Box margin="40px">
          <TableContainer component={Paper}>
            <Table sx={{ minWidth: 650 }} aria-label="event visitors table">
              <TableHead>
                <TableRow>
                  <TableCell sx={{ fontWeight: "bold" }} align="center">
                    Visitor ID
                  </TableCell>
                  <TableCell sx={{ fontWeight: "bold" }} align="center">
                    Full name
                  </TableCell>
                  <TableCell sx={{ fontWeight: "bold" }} align="center">
                    Email
                  </TableCell>
                  <TableCell sx={{ fontWeight: "bold" }} align="center">
                    Age
                  </TableCell>
                  <TableCell sx={{ fontWeight: "bold" }} align="center">
                    Date of birth
                  </TableCell>
                  <TableCell align="center"></TableCell>
                  <TableCell align="center"></TableCell>
                </TableRow>
              </TableHead>
              <TableBody>
                {eventVisitors.map((eventVisitor) => (
                  <EventVisitor
                    key={eventVisitor.id}
                    eventVisitor={eventVisitor}
                  />
                ))}
              </TableBody>
            </Table>
          </TableContainer>
        </Box>
      )}
    </Box>
  );
};
