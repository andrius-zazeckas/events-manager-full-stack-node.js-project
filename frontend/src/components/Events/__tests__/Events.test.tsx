import { BrowserRouter } from "react-router-dom";
import { Events } from "../Events";
import renderer from "react-test-renderer";

describe("Events", () => {
  it("matches snapshot", () => {
    const tree = renderer.create(
      <BrowserRouter>
        <Events />
      </BrowserRouter>
    );

    expect(tree).toMatchSnapshot();
  });
});
