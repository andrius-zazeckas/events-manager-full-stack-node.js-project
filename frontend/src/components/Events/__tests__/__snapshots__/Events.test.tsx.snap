// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Events matches snapshot 1`] = `
<div
  className="MuiBox-root css-1h31drq"
>
  <div
    className="MuiBox-root css-1qapnz0"
  >
    <h3
      className="MuiTypography-root MuiTypography-h3 css-gepadz-MuiTypography-root"
    >
      Loading...
    </h3>
  </div>
</div>
`;
